x/zoneconcierge/keeper/grpc_query.go

```diff
@@ -0,0 +1,403 @@
+
+	resp := &types.QueryEpochChainsInfoResponse{ChainsInfo: chainsInfo}
+	return resp, nil
+}
+
+// ListHeaders returns all headers of a chain with given ID, with pagination support
+func (k Keeper) ListHeaders(c context.Context, req *types.QueryListHeadersRequest) (*types.QueryListHeadersResponse, error) {
+	if req == nil {
+		return nil, status.Error(codes.InvalidArgument, "invalid request")
+	}
+
+	if len(req.ConsumerId) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "chain ID cannot be empty")
+	}
+
+	ctx := sdk.UnwrapSDKContext(c)
+
+	headers := []*types.IndexedHeader{}
+	store := k.canonicalChainStore(ctx, req.ConsumerId)
+	pageRes, err := query.Paginate(store, req.Pagination, func(key, value []byte) error {
+		var header types.IndexedHeader
+		k.cdc.MustUnmarshal(value, &header)
+		headers = append(headers, &header)
+		return nil
+	})
+	if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	resp := &types.QueryListHeadersResponse{
+		Headers:    headers,
+		Pagination: pageRes,
+	}
+	return resp, nil
+}
+
+// ListEpochHeaders returns all headers of a chain with given ID
+// TODO: support pagination in this RPC
+func (k Keeper) ListEpochHeaders(c context.Context, req *types.QueryListEpochHeadersRequest) (*types.QueryListEpochHeadersResponse, error) {
+	if req == nil {
+		return nil, status.Error(codes.InvalidArgument, "invalid request")
+	}
+
+	if len(req.ConsumerId) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "chain ID cannot be empty")
+	}
+
+	ctx := sdk.UnwrapSDKContext(c)
+
+	headers, err := k.GetEpochHeaders(ctx, req.ConsumerId, req.EpochNum)
+	if err != nil {
+		return nil, err
+	}
+
+	resp := &types.QueryListEpochHeadersResponse{
+		Headers: headers,
+	}
+	return resp, nil
+}
+
+// FinalizedChainsInfo returns the finalized info for a given list of chains
+func (k Keeper) FinalizedChainsInfo(c context.Context, req *types.QueryFinalizedChainsInfoRequest) (*types.QueryFinalizedChainsInfoResponse, error) {
+	if req == nil {
+		return nil, status.Error(codes.InvalidArgument, "invalid request")
+	}
+
+	// return if no chain IDs are provided
+	if len(req.ConsumerIds) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "chain ID cannot be empty")
+	}
+
+	// return if chain IDs exceed the limit
+	if len(req.ConsumerIds) > maxQueryChainsInfoLimit {
+		return nil, status.Errorf(codes.InvalidArgument, "cannot query more than %d chains", maxQueryChainsInfoLimit)
+	}
+
+	// return if chain IDs contain duplicates or empty strings
+	if err := bbntypes.CheckForDuplicatesAndEmptyStrings(req.ConsumerIds); err != nil {
+		return nil, status.Error(codes.InvalidArgument, types.ErrInvalidConsumerIDs.Wrap(err.Error()).Error())
+	}
+
+	ctx := sdk.UnwrapSDKContext(c)
+	resp := &types.QueryFinalizedChainsInfoResponse{FinalizedChainsInfo: []*types.FinalizedChainInfo{}}
+
+	// find the last finalised epoch
+	lastFinalizedEpoch := k.GetLastFinalizedEpoch(ctx)
+	for _, ConsumerId := range req.ConsumerIds {
```

Comment:
Thinking if doing all of this to 100 consumers at once wouldn't it be too much? 
Maybe we should only allow to query one by one

Reply:
Added a TODO for paginating this for loop. This is not pressiong though as we expect the integration to be permissioned in the beginning and there won't be too many of them